### PR TITLE
feat(tests): Add multiple exception support to EOF tests

### PR DESCRIPTION
**Pull request by shemnon**
## 🗒️ Description
There are some tests that are impossible to express without causing
multiple validation exceptions. Clients should be free to expose any of
those exceptions. To support this pipe notation for expected exceptions
is plumed through the EOF tests and implemented in some tests.

## ✅ Checklist
- [ ] All: Set appropriate labels for the changes.
- [ ] All: Considered squashing commits to improve commit history.
- [ ] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [ ] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [ ] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

### CODE REVIEW

**Comment by marioevz** [comment]
The `EOFException` base exception type already allows concatenating exceptions to allow the returned exception to be any of a given list, we could do that first because I'm worried that being too flexible is sometimes a bad thing when it comes to tests verification.

These concatenated exceptions are dumped into the fixtures in the form of a pipe-separated string:
E.g.
```python
EOFException.INVALID_VERSION | EOFException.UNDEFINED_INSTRUCTION
```
is dumped into the fixture as:
```json
"results": {
    "Prague": {
        "exception": "EOFException.INVALID_VERSION|EOFException.UNDEFINED_INSTRUCTION",
        "result": false
    }
}
```

Could this apply to the cases included in this PR?

**Comment by shemnon** [comment]
I'll definitely look into this.  There are many more error cases than are seen in this PR, particualrly in he area of invalid laouts.

**Comment by shemnon** [comment]
The code using EOFExceptions was not expecting the instance or list nomenclature, so I re-booted the PR, and plumbed exception instance or list support through eof_test and updated all the tests Besu will report multiple exceptions for.

File: src/ethereum_test_specs/eof.py [hunk 1]
```diff
@@ -237,22 +238,43 @@ def verify_result(self, result: CompletedProcess, expected_result: Result, code:
...
-        expected_exception = expected_result.exception
-        expected_message = parser.exception_to_message(expected_exception)
-
-        if "OK" in actual_message:
-            raise ExpectedEOFException(
-                code=code, expected=f"{expected_exception} ({expected_message})"
-            )
+            for expected in expected_result.exception:
+                if expected == actual_exception:
+                    return
```
Comment (marioevz) [inline src/ethereum_test_specs/eof.py:252]:
```suggestion
            if actual_exception in expected_result.exception:
                return
```
This should do the trick, and it also works for non-list exceptions.

**Review by marioevz — COMMENTED** [review]
Changes look good so far, just a couple of minor comments. Thanks!

File: src/ethereum_test_specs/eof.py [hunk 2]
```diff
@@ -237,22 +238,43 @@ def verify_result(self, result: CompletedProcess, expected_result: Result, code:
...
 
+        else:
+            expected_exception = expected_result.exception
+            expected_message = parser.exception_to_message(expected_exception)
+
+            if "OK" in actual_message:
+                raise ExpectedEOFException(
+                    code=code,
+                    expected=f"{expected_exception} ({expected_message})",
+                    got="balid EOF code",
```
Comment (marioevz) [inline src/ethereum_test_specs/eof.py:268]:
```suggestion
```

File: src/ethereum_test_specs/eof.py [hunk 3]
```diff
@@ -237,22 +238,43 @@ def verify_result(self, result: CompletedProcess, expected_result: Result, code:
                 raise UnexpectedEOFException(
                     code=code, got=f"{actual_exception} ({actual_message})"
                 )
+        elif isinstance(expected_result.exception, List):
+            expected_string = to_pipe_str(expected_result.exception)
+            if "OK" in actual_message:
+                raise ExpectedEOFException(
+                    code=code,
+                    expected=f"multiple possible exceptions {expected_string}",
+                    got="valid EOF code",
```
Comment (marioevz) [inline src/ethereum_test_specs/eof.py:247]:
```suggestion
```

**Review by marioevz — COMMENTED** [review]
Couple more comments to make tox happy.

File: src/ethereum_test_specs/eof.py [hunk 4]
```diff
@@ -13,6 +13,7 @@
 
 from ethereum_test_base_types import Account, Bytes
 from ethereum_test_exceptions import EOFException, EvmoneExceptionMapper
+from ethereum_test_exceptions.exceptions import EOFExceptionInstanceOrList, to_pipe_str
```
[the remark this answers is not in the document]
Comment (shemnon) [reply src/ethereum_test_specs/eof.py:16]: Done

File: src/ethereum_test_specs/eof.py (hunk 1 above)
Comment (shemnon) [reply src/ethereum_test_specs/eof.py]: Done, and dropped it back to one case for single or multiple exceptions

File: src/ethereum_test_specs/eof.py (hunk 2 above)
Comment (shemnon) [reply src/ethereum_test_specs/eof.py]: done

File: src/ethereum_test_specs/eof.py (hunk 3 above)
Comment (shemnon) [reply src/ethereum_test_specs/eof.py]: Moot, whole section removed.

**Review by marioevz — APPROVED** [review]
LGTM, I'll push a commit containing the changelog update and then will merge.

**Pull request merged**